File: frontend/src/lib/navhistory.js
// Back / forward memory for one file pane - the browser behaviour: visiting a
// directory pushes it, Back walks towards older entries without forgetting the
// newer ones, and visiting something new while parked in the middle drops the
// forward tail. Mirrors mysql_runner/transfer/navhistory.py.

const LIMIT = 200

export class NavHistory {
  #entries = []
  #index = -1
  #limit

  constructor(limit = LIMIT) {
    this.#limit = Math.max(2, limit)
  }

  get current() {
    return this.#index >= 0 ? this.#entries[this.#index] : ''
  }

  get entries() {
    return [...this.#entries]
  }

  get index() {
    return this.#index
  }

  get canGoBack() {
    return this.#index > 0
  }

  get canGoForward() {
    return this.#index > -1 && this.#index < this.#entries.length - 1
  }

  /** Record a directory the user navigated to. A refresh is not a navigation. */
  visit(path) {
    if (!path || this.current === path) return
    this.#entries.splice(this.#index + 1)
    this.#entries.push(path)
    if (this.#entries.length > this.#limit) {
      this.#entries.splice(0, this.#entries.length - this.#limit)
    }
    this.#index = this.#entries.length - 1
  }

  back() {
    if (!this.canGoBack) return ''
    this.#index -= 1
    return this.#entries[this.#index]
  }

  forward() {
    if (!this.canGoForward) return ''
    this.#index += 1
    return this.#entries[this.#index]
  }

  go(index) {
    if (index < 0 || index >= this.#entries.length) return ''
    this.#index = index
    return this.#entries[this.#index]
  }

  /** Most recently visited directories, newest first, without duplicates. */
  recent(count = 10) {
    const seen = []
    for (let i = this.#index; i >= 0; i -= 1) {
      const path = this.#entries[i]
      if (!seen.includes(path)) seen.push(path)
      if (seen.length >= count) break
    }
    return seen
  }
}

/**
 * Translate a path under one base into the matching path under another - what
 * mirrored navigation needs. Returns '' when the path is not inside the base.
 */
export function mirrorPath(sourceBase, sourcePath, targetBase, { posix }) {
  const rel = relative(sourceBase, sourcePath)
  if (rel === null) return ''
  if (!rel) return targetBase
  if (posix) {
    return (targetBase === '/' ? '' : targetBase.replace(/\/+$/, '')) + '/' + rel
  }
  const separator = targetBase.includes('/') && !targetBase.includes('\\') ? '/' : '\\'
  return targetBase.replace(/[\\/]+$/, '') + separator + rel.split('/').join(separator)
}

function split(path) {
  return path
    .replace(/\\/g, '/')
    .split('/')
    .filter((part) => part !== '' && part !== '.')
}

function relative(base, path) {
  const baseParts = split(base)
  const pathParts = split(path)
  if (pathParts.length < baseParts.length) return null
  for (let i = 0; i < baseParts.length; i += 1) {
    // Windows paths are case-insensitive, and comparing them as given would
    // break mirroring the moment a drive letter's case differed.
    if (baseParts[i].toLowerCase() !== pathParts[i].toLowerCase()) return null
  }
  return pathParts.slice(baseParts.length).join('/')
}

File: mysql_runner/transfer/navhistory.py
"""Back / forward navigation memory for a file pane.

A browser-style stack: visiting a directory pushes it, Back walks towards
older entries without forgetting the newer ones, and visiting something new
while parked in the middle discards the forward tail. Kept free of Qt so it can
be tested on its own and reused by both front ends.
"""

from __future__ import annotations

import posixpath

#: How many directories to remember per pane.
LIMIT = 200


class NavHistory:
    """The visited-directory stack for one pane."""

    def __init__(self, limit: int = LIMIT) -> None:
        self._entries: list[str] = []
        self._index = -1
        self._limit = max(2, limit)

    # ----- state ----------------------------------------------------------
    @property
    def current(self) -> str:
        if 0 <= self._index < len(self._entries):
            return self._entries[self._index]
        return ""

    @property
    def entries(self) -> list[str]:
        return list(self._entries)

    @property
    def index(self) -> int:
        return self._index

    def can_go_back(self) -> bool:
        return self._index > 0

    def can_go_forward(self) -> bool:
        return -1 < self._index < len(self._entries) - 1

    def recent(self, count: int = 10) -> list[str]:
        """The most recently visited directories, newest first, deduplicated."""
        seen: list[str] = []
        for path in reversed(self._entries[: self._index + 1]):
            if path not in seen:
                seen.append(path)
            if len(seen) >= count:
                break
        return seen

    # ----- movement -------------------------------------------------------
    def visit(self, path: str) -> None:
        """Record a directory the user navigated to."""
        if not path:
            return
        if self.current == path:
            return  # A refresh is not a navigation.
        del self._entries[self._index + 1:]
        self._entries.append(path)
        if len(self._entries) > self._limit:
            # Drop the oldest; the index moves with it.
            overflow = len(self._entries) - self._limit
            del self._entries[:overflow]
        self._index = len(self._entries) - 1

    def back(self) -> str:
        """Step back one entry and return it ("" when there is none)."""
        if not self.can_go_back():
            return ""
        self._index -= 1
        return self._entries[self._index]

    def forward(self) -> str:
        """Step forward one entry and return it ("" when there is none)."""
        if not self.can_go_forward():
            return ""
        self._index += 1
        return self._entries[self._index]

    def go(self, index: int) -> str:
        """Jump to an absolute position in the stack (for a drop-down menu)."""
        if not 0 <= index < len(self._entries):
            return ""
        self._index = index
        return self._entries[self._index]

    def clear(self) -> None:
        self._entries.clear()
        self._index = -1


def mirror_path(source_base: str, source_path: str, target_base: str, *, posix: bool) -> str:
    """Translate ``source_path`` under ``source_base`` into ``target_base``.

    This is what "mirror navigation" needs: the user walks into
    ``/var/www/site/public`` on one side and the other side should follow to
    ``…/site/public`` beneath its own root. Returns "" when ``source_path`` is
    not inside ``source_base``, which is the caller's cue to leave the other
    pane alone.
    """
    source_rel = _relative(source_base, source_path)
    if source_rel is None:
        return ""
    if not source_rel:
        return target_base
    if posix:
        return posixpath.normpath(posixpath.join(target_base or "/", source_rel))
    import os  # local: keeps the POSIX path of this module import-light

    return os.path.normpath(os.path.join(target_base, source_rel.replace("/", os.sep)))


def _relative(base: str, path: str) -> str | None:
    """POSIX-ish relative path, tolerating both separators. None if outside."""
    base_parts = _split(base)
    path_parts = _split(path)
    if len(path_parts) < len(base_parts):
        return None
    for left, right in zip(base_parts, path_parts):
        if left.lower() != right.lower():  # Windows paths are case-insensitive
            return None
    return "/".join(path_parts[len(base_parts):])


def _split(path: str) -> list[str]:
    return [part for part in path.replace("\\", "/").split("/") if part not in ("", ".")]
